File: 0784-Letter Case Permutation/0784-Letter Case Permutation.cpp
class Solution {
public:
    vector<string> letterCasePermutation(string S) {
        vector<string> result;
        string path;
        dfs(S, 0, path, result);
        return result;
    }
    
private:
    void dfs(string& s, int start, string& path, vector<string>& result) {
        if (start == s.size()) {
            result.push_back(path);
            return;
        }
        
        if (isdigit(s[start])) {
            path += s[start];
            dfs(s, start + 1, path, result);
            path.pop_back();
        }
        else {
            path += tolower(s[start]);
            dfs(s, start + 1, path, result);
            path.pop_back();
            
            path += toupper(s[start]);
            dfs(s, start + 1, path, result);
            path.pop_back();
        }
    }
};

class Solution {
public:
    vector<string> letterCasePermutation(string S) {
        vector<string> result;
        int count = 0;
        for (char c : S) {
            if (!isdigit(c)) ++count;
        }
        int upper = 1 << count;
        for (int i = 0; i < upper; ++i) {
            string s;
            int b = 0;
            for (char c : S) {
                if (isdigit(c)) {
                    s += c;
                }
                else {
                    if ((i >> b++) & 1) {
                        s += toupper(c);
                    }
                    else {
                        s += tolower(c);
                    }
                }
            }
            result.push_back(s);
        }
        return result;
    }
};

File: 0701-0800/0784-Letter Case Permutation/0784-Letter Case Permutation.py
class Solution:
    def letterCasePermutation(self, S: str) -> List[str]:
        count = sum(c.isalpha() for c in S)
        result = []
        for i in range(1 << count):
            b = 0
            curr = []
            for c in S:
                if c.isdigit():
                    curr.append(c)
                else:
                    if (1 << b) & i:
                        curr.append(c.upper())
                    else:
                        curr.append(c.lower())
                    b += 1
            result.append(''.join(curr))
        return result
